src/shapirowilk.jl

```diff
@@ -0,0 +1,244 @@
+end
+
+struct ShapiroWilkTest <: HypothesisTest
+    coefs::ShapiroWilkCoefs # Expectation of order statistics for Shapiro-Wilk test
+    W::Float64              # test statistic
+    censored::Int   # only the smallest N₁ samples were used
+end
+
+testname(::ShapiroWilkTest) = "Shapiro-Wilk normality test"
+function population_param_of_interest(t::ShapiroWilkTest)
+    return ("Squared correlation of data and the expected order statistics of N(0,1) (W)",
+            1.0, t.W)
+end
+default_tail(::ShapiroWilkTest) = :left
+censored_ratio(t::ShapiroWilkTest) = t.censored / length(t.coefs)
+
+function show_params(io::IO, t::ShapiroWilkTest, indent)
+    l = 24
+    println(io, indent, rpad("number of observations:", l), length(t.coefs))
+    println(io, indent, rpad("censored ratio:", l), censored_ratio(t))
+    return println(io, indent, rpad("W-statistic:", l), t.W)
+end
+
+function StatsAPI.pvalue(t::ShapiroWilkTest)
+    n = length(t.coefs)
+    W = t.W
+
+    if iszero(censored_ratio(t))
+        if n == 3 # exact by Shapiro&Wilk 1965
+            # equivalent to 6/π * (asin(sqrt(W)) - asin(sqrt(3/4)))
+            return 1 - 6acos(sqrt(W)) / π
+        elseif n ≤ 11 # Royston 1992
+            γ = __RS92_G(n)
+            if γ ≤ log1p(-W)
+                return zero(W)
+            end
+            w = -log(γ - log1p(-W))
+            μ = __RS92_C3(n)
+            σ = exp(__RS92_C4(n))
+        elseif 12 ≤ n # Royston 1992
+            w = log1p(-W)
+            μ = __RS92_C5(log(n))
+            σ = exp(__RS92_C6(log(n)))
+        end
+        return ccdf(Normal(μ, σ), w)
+    else
+        throw("censored samples not implemented yet")
+        # to implement censored samples follow Royston 1993 Section 3.3
+    end
+end
+
+"""
+    ShapiroWilkTest(X::AbstractArray{<:Real},
+                    swc::ShapiroWilkCoefs=ShapiroWilkCoefs(length(X));
+                    sorted::Bool=issorted(X),
+                    censored::Integer=0)
+
+Perform a Shapiro-Wilk test of the null hypothesis that the data in array `X`
+come from a normal distribution.
+
+This implementation is based the method by Royston (1992).
+The calculation of the p-value is exact for `N = 3`, and for ranges
+`4 ≤ N ≤ 11` and `12 ≤ N ≤ 5000` (Royston 1992) two separate approximations
+for p-values are used.
+
+# Keyword arguments
+The following keyword arguments may be passed.
+ * `sorted::Bool=issorted(X)`: to indicate that data `X` is already sorted.
+ * `censored::Integer=0`: to censor the largest samples from `X` (so called upper-tail censoring)
+   Note: currently `pvalue` is not implemented for censored samples.
+
+Implements: [`pvalue`](@ref)
+
+# Warning
+As noted by Royston (1993), (approximated) W-statistic will be accurate
+but returned p-values may not be reliable if either of these apply:
+* Sample size is large  (`N > 2000`) or small (`N < 20`)
+* Too much data is censored (`censored / N > 0.8`)
+
+# Implementation notes
```

Comment:
These are not really implementation notes except for the first point. And conversely, a paragraph above is about the implementation but it's not under this section.

The docstring should explain how to create an object to pass to `swc`, and it should be mentioned earlier as it's hard to find what the argument means currently.

Reply:
hmm, I actually see it differently -- the paragraph above is about the method (described by Royston). Implementation notes are about this particular implementation of the method.

If you feel like a change in headers is necessary, let me know.

Reply:
This is not how docstrings are usually organized in Julia packages AFAICT. And I'm not sure it's really useful for users to separate the description of the method (that they often won't know about) and the Julia implementation.